### PR TITLE
[tests] Migrate from yarn to pnpm

### PR DESCRIPTION
**Pull request by Ethan-Arrowood**
<picture data-single-emoji=":pnpm:" title=":pnpm:">:pnpm:</picture> 

yarn has become increasingly more difficult to use as the v1 we rely on no longer receives updates. pnpm is faster and is actively maintained. 

This PR migrates us to pnpm.

### CODE REVIEW

File: package.json
Comment (TooTallNate) [inline package.json:7]: We don't need this anymore? I think perhaps lerna relies on it.

File: packages/cli/package.json
Comment (TooTallNate) [inline packages/cli/package.json]: What's this?

Comment (Ethan-Arrowood) [reply packages/cli/package.json]: that looks like a mistake. i was running a lot of `pnpm i` commands to get everything building

File: package.json
Comment (Ethan-Arrowood) [reply package.json:7]: not sure - lerna is used for publishing packages only right? I can keep this in here, but it seemed like they were `yarn` configurations. I wonder if since I configured lerna to `pnpm` it will look at the `pnpm-workspace.yaml` instead

File: packages/cli/package.json
Comment (TooTallNate) [inline packages/cli/package.json]: Is `workspace:` necessary here?

File: packages/cli/package.json
Comment (cb1kenobi) [reply packages/cli/package.json]: Ha, I've done this before :)

File: packages/cli/package.json
Comment (Ethan-Arrowood) [reply packages/cli/package.json]: Ahh thats a pnpm thing. https://pnpm.io/cli/add#install-from-the-workspace

Might have to update all them to use this `workspace:` syntax

File: package.json
Comment (cb1kenobi) [reply package.json:7]: @TooTallNate is correct, `lerna` still needs `workspaces` in the `package.json`.

```
$ lerna changed
lerna notice cli v6.0.3
lerna ERR! EWORKSPACES Workspaces need to be defined in the root package.json.
lerna ERR! EWORKSPACES See: https://github.com/lerna/lerna/blob/master/commands/bootstrap/README.md#--use-workspaces
```

File: .github/workflows/publish.yml
Comment (styfle) [inline .github/workflows/publish.yml]: Is this whitespace change correct? Curious why its changing now?

Comment (Ethan-Arrowood) [reply .github/workflows/publish.yml]: Might be some auto formatting on my end - will revert

File: packages/cli/package.json
Comment (cb1kenobi) [reply packages/cli/package.json]: Related/unrelated, FWIW when you create a new Turbo repo, the template uses `workspace:`: https://github.com/vercel/turbo/blob/main/examples/basic/packages/ui/package.json#L14.

File: .github/workflows/publish.yml
Comment (styfle) [inline .github/workflows/publish.yml]: Before each `pnpm install` step, there should be a `corepack enable` step.

File: .github/workflows/publish.yml
Comment (Ethan-Arrowood) [reply .github/workflows/publish.yml]: Unless its a correct fix? I trust the auto formatter more than myself lol tho yaml is yaml so 🤷‍♂️

Comment (styfle) [reply .github/workflows/publish.yml]: I'm guessing you have a vs code plugin thats formatting and its not something our repo automatically does

File: packages/cli/package.json
Comment (styfle) [reply packages/cli/package.json]: Yeah that seems correct for `turbo`. However, does `lerna` know how to handle `workspace:`?

File: packages/gatsby-plugin-vercel-analytics/README.md
Comment (styfle) [inline packages/gatsby-plugin-vercel-analytics/README.md]: Lets either add all options (npm/yarn/pnpm) to the readme, or just have npm only

File: packages/next/package.json
Comment (styfle) [inline packages/next/package.json]: Seems like the `save-prefix: ""` setting didn't work because the up-caret is still there.

Perhaps we should use `save-exact: true` instead.

File: packages/redwood/package.json
Comment (styfle) [inline packages/redwood/package.json]: Same here: unexpected `^`

File: utils/publish.sh
Comment (styfle) [inline utils/publish.sh:33]: I think we can remove this checkout now since pnpm defaults to frozen lockfile in CI, so it should never happen.

Although we can circle back to it after this PR lands.

File: packages/cli/package.json
Comment (Ethan-Arrowood) [reply packages/cli/package.json]: https://lerna.js.org/docs/recipes/using-pnpm-with-lerna it supports it

File: package.json
Comment (Ethan-Arrowood) [reply package.json:7]: It seems lerna _should_ be able to use pnpm-workspace.yaml https://lerna.js.org/docs/recipes/using-pnpm-with-lerna investigating now

Comment (Ethan-Arrowood) [reply package.json:7]: Ahh it must only support it in like v5+ - we are using v3 :(

File: .github/workflows/publish.yml
Comment (Ethan-Arrowood) [reply .github/workflows/publish.yml]: Yeah, but my other comment still stands. Is this change okay if its _better_ yaml format? Other workflows have this indentation. @andymckay do you know if this indentation is okay to keep or do I need to revert and disable my auto-formatter for yaml files?

File: package.json
Comment (cb1kenobi) [reply package.json:7]: Can we update Lerna? The last v3 release was 2-1/2 years ago. Lerna (again) is being maintained and it makes most sense to keep current. I'm not sure how the release process would be affected.

Comment (Ethan-Arrowood) [reply package.json:7]: imo we should migrate off lerna. use pnpm+turbo instead

Comment (Ethan-Arrowood) [reply package.json:7]: I can try updating it now to atleast v5 

v6 uses Nx lol

Comment (cb1kenobi) [reply package.json:7]: Sigh, figures. +1 for pnpm+turbo. Are there any projects out there that uses pnpm+turbo for publishing/changelog that we can look to for inspiration?

Comment (Ethan-Arrowood) [reply package.json:7]: https://github.com/vercel/turbo lol they've got a whole changeset workflow we can clone

Comment (Ethan-Arrowood) [reply package.json:7]: oh nvm they moved to something else: https://github.com/vercel/turbo/blob/main/.github/workflows/publish-npm.yml

but we should still consult them.

Comment (Ethan-Arrowood) [reply package.json:7]: yay lerna5 works! try it out locally and let me know if you hit some edge case I didn't

File: packages/cli/package.json
Comment (Ethan-Arrowood) [reply packages/cli/package.json]: It does in v5! Migrating all workspace packages to the prefix

File: .github/workflows/test-unit.yml
Comment (Ethan-Arrowood) [inline .github/workflows/test-unit.yml]: @styfle this spacing fix seems correct. If anything we probably need to be enforcing format on these files

File: .github/workflows/publish.yml
Comment (Ethan-Arrowood) [reply .github/workflows/publish.yml]: If you compare this spacing change to the other files you can see how in the `test-*.yml` files they have indents after `steps:`, but its in this file they are all inline with `steps:`.

File: packages/next/package.json
Comment (Ethan-Arrowood) [reply packages/next/package.json]: save-exact is not supported by pnpm so just adding save-prefix which is

File: utils/publish.sh
Comment (Ethan-Arrowood) [reply utils/publish.sh:33]: lets circle back after PR

File: packages/next/package.json
Comment (styfle) [reply packages/next/package.json]: Oddly enough, `save-exact=true` work for me with pnpm even though its not documented

Comment (Ethan-Arrowood) [reply packages/next/package.json]: are you putting that in `.npmrc` ? I tried it and it didn't work for some reason

Comment (styfle) [reply packages/next/package.json]: Yes in `.npmrc` with `pnpm@7.22.0`

https://github.com/styfle/links-awakening/blob/main/.npmrc

Oddly enough, I couldn't get `save-prefix` to work when I tried it last.

File: .github/workflows/test-unit.yml
Comment (styfle) [reply .github/workflows/test-unit.yml]: I would prefer the spacing is fixed in a different PR if possible and added to prettier (since it seems like its currently not auto-formatted)

The reason being: this PR is changing a lot so we should reduce the lines touched where possible

File: packages/next/package.json
Comment (Ethan-Arrowood) [reply packages/next/package.json]: lol really? that is interesting. let me try again

File: .github/workflows/test-unit.yml
Comment (Ethan-Arrowood) [reply .github/workflows/test-unit.yml]: Fair enough - ill revert and make sure to follow up

File: packages/next/package.json
Comment (Ethan-Arrowood) [reply packages/next/package.json]: okay both seem to work. ill stick with save-exact=true

File: packages/python/package.json
Comment (styfle) [inline packages/python/package.json]: This should be in `devDependencies`

File: .github/workflows/publish.yml
Comment (Ethan-Arrowood) [reply .github/workflows/publish.yml]: @TooTallNate and I think that the `corepack enabled` is now disabling yarn from executing within the CLI and causing some tests to fail. Ideas/thoughts? Maybe we don't use corepack and instead just install pnpm manually or use [pnpm/action-setup](https://github.com/pnpm/action-setup) ?

**Comment by styfle** [comment]
In the future we, should use `COREPACK_ENABLE_STRICT=0` however that version of corepack with this feature hasn't shipped in Node.js 14 yet. So might as well install pnpm directly.

- https://github.com/nodejs/corepack/issues/195

File: .github/workflows/test.yml
Comment (styfle) [inline .github/workflows/test.yml]: Lets make sure to use `7.24.2` for all our installs so it matches what corepack would have used

File: packages/build-utils/test/fixtures/15-npm-8-legacy-peer-deps/package-lock.json
Comment (Ethan-Arrowood) [inline packages/build-utils/test/fixtures/15-npm-8-legacy-peer-deps/package-lock.json:12]: @styfle you were the last to touch this one. Not sure why this is getting generated but wasn't before. I've tried reverting, but it generates when I run the unit tests locally.

File: packages/build-utils/test/unit.test.ts
Comment (Ethan-Arrowood) [inline packages/build-utils/test/unit.test.ts]: @TooTallNate not sure if theres a better way to do this, but these paths work and now require the project being build first since it can't rely on hoisting lol

File: packages/build-utils/test/unit.walk.test.ts
Comment (Ethan-Arrowood) [inline packages/build-utils/test/unit.walk.test.ts:5]: This was a weird one, was getting this error that seemed only solvable by doing this change. Must be connected to finally using `@types/ndoe@14`

```
Assertions require every name in the call target to be declared with an explicit type annotation.ts(2775)
unit.walk.test.ts(5, 9): 'deepEqual' needs an explicit type annotation.
```

File: packages/edge/docs/README.md
Comment (Ethan-Arrowood) [inline packages/edge/docs/README.md:39]: Not sure why these had to change. But the tests were failing and said to re run the docs:build script 🤷‍♂️

File: packages/build-utils/test/fixtures/15-npm-8-legacy-peer-deps/package-lock.json
Comment (styfle) [reply packages/build-utils/test/fixtures/15-npm-8-legacy-peer-deps/package-lock.json:12]: Its probably fine if the tests pass.

The test is checking to see if `npm install` fails so I'm a little surprised if the lock file updates

See https://github.com/vercel/vercel/pull/8598

File: packages/build-utils/test/unit.test.ts
Comment (styfle) [reply packages/build-utils/test/unit.test.ts]: This test was explicitly designed to check the package boundary since we used to have builders that did this:

```
require('@vercel/build-utils/fs/get-writable-directory.js');
```

File: packages/build-utils/test/unit.walk.test.ts
Comment (styfle) [reply packages/build-utils/test/unit.walk.test.ts:5]: Lets bump the types then?

File: packages/edge/docs/interfaces/ExtraResponseInit.md
Comment (styfle) [inline packages/edge/docs/interfaces/ExtraResponseInit.md]: Might want to get approval from the edge team on this change since this looks unexpected

File: packages/build-utils/test/unit.walk.test.ts
Comment (Ethan-Arrowood) [reply packages/build-utils/test/unit.walk.test.ts:5]: If i'm not mistaken bumping types to 16 causes a bunch of other issues. the whole repo is on 14 so we should keep it that way

File: api/_lib/util/error-handler.ts
Comment (EndangeredMassa) [inline api/_lib/util/error-handler.ts:17]: **suggestion (non-blocking):** add comment about why

Can you add a comment here about why this was ignored?

File: packages/build-utils/src/fs/run-user-scripts.ts
Comment (EndangeredMassa) [inline packages/build-utils/src/fs/run-user-scripts.ts:142]: **suggestion (non-blocking):** add comment about why

Can you add a comment here about why this was ignored?

File: packages/build-utils/package.json
Comment (TooTallNate) [inline packages/build-utils/package.json]: Why was the `turbo build` part necessary?

File: api/package.json
Comment (EndangeredMassa) [inline api/package.json:18]: **question (non-blocking):** Is everything supposed to be node@14 or no?

https://github.com/vercel/vercel/pull/9198/files#r1067359556

File: packages/build-utils/test/unit.test.ts
Comment (TooTallNate) [reply packages/build-utils/test/unit.test.ts]: Maybe this is slightly better. Personally I wouldn't mind deleting this test + legacy behavior entirely, but idk might be too risky.

```suggestion
  const index = require('../');
```

File: packages/build-utils/package.json
Comment (Ethan-Arrowood) [reply packages/build-utils/package.json]: Since these tests now require the built module, we must ensure its being built every time tests are run. turbo here lets that build hit the cache if nothing has changed. Can change to just `pnpm build` but then no caching will happen

File: packages/build-utils/test/unit.walk.test.ts
Comment (TooTallNate) [reply packages/build-utils/test/unit.walk.test.ts:5]: I'd prefer to switch to `expect`

File: packages/cli/package.json
Comment (TooTallNate) [reply packages/cli/package.json]: I'm curious how this works with the published package. Does lerna change the version numbers when publishing?

File: packages/client/README.md
Comment (EndangeredMassa) [inline packages/client/README.md:14]: **question (non-blocking):** why was this removed?

File: packages/build-utils/test/unit.walk.test.ts
Comment (Ethan-Arrowood) [reply packages/build-utils/test/unit.walk.test.ts:5]: I consider that out of scope for this PR. We can update tests in packages in other PRs since that is its own effort

File: packages/edge/scripts/fix-links.js
Comment (EndangeredMassa) [inline packages/edge/scripts/fix-links.js:9]: **suggestion (non-blocking):** add comment about why this is necessary

Can we add a comment here about what exactly is happening and why? It could link to the issue you created: https://github.com/TypeStrong/typedoc/issues/2140

File: packages/gatsby-plugin-vercel-analytics/README.md
Comment (EndangeredMassa) [inline packages/gatsby-plugin-vercel-analytics/README.md:26]: **question (non-blocking):** why was this removed?

File: utils/chunk-tests.js
Comment (EndangeredMassa) [inline utils/chunk-tests.js]: **suggestion (non-blocking):** remove?

```suggestion
```

Is this necessary? If it is, can we add a comment about why?

**Comment by EndangeredMassa** [comment]
I tested this out locally and it seems to work pretty well. 🎉

File: packages/gatsby-plugin-vercel-analytics/README.md
Comment (Ethan-Arrowood) [reply packages/gatsby-plugin-vercel-analytics/README.md:26]: @styfle previous comment about our docs should either have `yarn/npm/pnpm` or just `npm`

File: packages/build-utils/package.json
Comment (styfle) [reply packages/build-utils/package.json]: @Ethan-Arrowood Something doesn't seem right.

I would expect `turbo build` to automatically be run when you do `turbo test-unit` because of our config here:

https://github.com/vercel/vercel/blob/3f47587a8b204ecc5782fde77dabcee5f1d6e7e7/turbo.json#L24-L25

File: utils/chunk-tests.js
Comment (Ethan-Arrowood) [reply utils/chunk-tests.js]: necessary because how  pnpm handles argument parsing

File: packages/client/README.md
Comment (Ethan-Arrowood) [reply packages/client/README.md:14]: @styfle previous comment about our docs should either have yarn/npm/pnpm or just npm

File: api/package.json
Comment (Ethan-Arrowood) [reply api/package.json:18]: yeah but not api since we deploy this with Node 16

File: packages/client/README.md
Comment (styfle) [reply packages/client/README.md:14]: Probably because of my comment here https://github.com/vercel/vercel/pull/9198#discussion_r1066224797

In hindsight, we should leave revert the changes to `packages/*/README.md` because they are for the consumer and therefore not relevant to this PR. They can be updated in a future PR.

File: packages/gatsby-plugin-vercel-analytics/README.md
Comment (styfle) [reply packages/gatsby-plugin-vercel-analytics/README.md:26]: Probably because of my comment here https://github.com/vercel/vercel/pull/9198#discussion_r1066224797

In hindsight, we should leave revert the changes to `packages/*/README.md` because they are for the consumer and therefore not relevant to this PR. They can be updated in a future PR.

File: packages/build-utils/package.json
Comment (Ethan-Arrowood) [reply packages/build-utils/package.json]: <picture data-single-emoji=":hmm:" title=":hmm:">:hmm:</picture>

Comment (Ethan-Arrowood) [reply packages/build-utils/package.json]: Yeah so this does work when `turbo` is used to invoke `test-unit`, but if you call `pnpm test-unit` it wont run

Comment (styfle) [reply packages/build-utils/package.json]: Yeah but thats expected. You should run `pnpm build` and then `pnpm test-unit` if you're running from the current directory without turbo.

File: packages/build-utils/package.json
Comment (styfle) [inline packages/build-utils/package.json:17]: This is wrong. It should be only running the integration tests, not the unit tests

Comment (Ethan-Arrowood) [reply packages/build-utils/package.json:17]: Huh? It is only running integration tests

Comment (styfle) [reply packages/build-utils/package.json:17]: Oh you're right. I misread that as `pnpm test-unit`. Ignore this comment 🙈

File: api/package.json
Comment (styfle) [inline api/package.json]:
```suggestion
```

You already did this TODO so we can remove it 🙂

File: utils/pack.ts
Comment (styfle) [inline utils/pack.ts:26]: Lets add a comment here explaining why

I think its because turbo is finding this by parsing the new pnpm-workspace.yml file but we don't want to create a tarball for it since its a Vercel Deployment already?

File: packages/cli/package.json
Comment (styfle) [inline packages/cli/package.json]: I think `@vercel/routing-utils` should be in devDependencies right? Since it wasn't there before?

File: packages/cli/package.json
Comment (styfle) [inline packages/cli/package.json]: Where is `qs` used?

File: packages/cli/package.json
Comment (styfle) [inline packages/cli/package.json]:
```suggestion
```

@Ethan-Arrowood Pretty sure we just removed `update-notifier`. 

cc @cb1kenobi

File: packages/edge/package.json
Comment (styfle) [inline packages/edge/package.json]: Any reason why these deps are updating?

File: packages/cli/package.json
Comment (Ethan-Arrowood) [reply packages/cli/package.json]: https://github.com/vercel/vercel/blob/main/packages/cli/src/util/git/connect-git-provider.ts#L2

File: packages/edge/package.json
Comment (Ethan-Arrowood) [reply packages/edge/package.json]: Trying to fix the edge docs issue. I can revert now that I have the replacer script

File: packages/next/package.json
Comment (styfle) [inline packages/next/package.json:13]: Why was `useStderr` removed?

Comment (Ethan-Arrowood) [reply packages/next/package.json:13]: these tests get passed `-- --listTests` by the chunk script and thus can't have any other arguments already in there. they seem to still run and pass fine without it so we went with it

File: packages/cli/package.json
Comment (styfle) [reply packages/cli/package.json]: :oof: ok we need to remove that in a follow up PR

**Review by EndangeredMassa — APPROVED** [review]

File: package.json
Comment (styfle) [inline package.json:8]: Any reason why this version of lerna was chosen?

The latest is 6.4.0 https://www.npmjs.com/package/lerna

Comment (Ethan-Arrowood) [reply package.json:8]: 6 uses Nx lol didn't want to muddy with Turbo. But we need a different major than 3 for `workspace:` support

File: package.json
Comment (styfle) [inline package.json]: I think `@vercel/fun` should be in `packages/cli/package.json` instead

File: package.json
Comment (styfle) [inline package.json:16]: Curious why `@vercel/next` is here. Where is it used?

File: package.json
Comment (Ethan-Arrowood) [reply package.json]: This chunk somehow needs it at the root level: 

https://github.com/vercel/vercel/blob/4c3bc0532216f21f354623ada76b2315ef8ba519/packages/cli/scripts/build.ts#L69-L76

(Remove, run install, run build, and you'll see the full error)

File: package.json
Comment (Ethan-Arrowood) [reply package.json:16]: Ahh this ones tricky. The `@vercel/next` test suite invokes the `test/lib/run-build-lambda.js` which does some funky `require(build.use)` call. `build.use` returns `'@vercel/next'` and so it has to be included in the root devDep list

**Review by TooTallNate — APPROVED** [review]

File: package.json
Comment (styfle) [reply package.json]: @Ethan-Arrowood Pretty sure that path was written that way because of yarn hoisting. So you can move `@vercel/fun` to the cli package and then update the `build.ts` to read the new path from `./node_modules/@vercel/fun/dist/src/runtimes` I believe.

@TooTallNate Do you agree?

File: package.json
Comment (styfle) [reply package.json:16]: Oh interesting. Perhaps we should have this script change directory into the package running it or update `use` to be the path to the current package.

@ijjk Do you have thoughts on this one?

https://github.com/vercel/vercel/blob/5b88f673f89c44b9e91e44247e29be6bed43805a/test/lib/run-build-lambda.js#L35

File: package.json
Comment (Ethan-Arrowood) [reply package.json]: ill give that a shot

Comment (Ethan-Arrowood) [reply package.json]: seems to build fine!

File: package.json
Comment (Ethan-Arrowood) [reply package.json:16]: this modification feels like scope creep to me. its more of a "update `@vercel/next` tests" kinda change

Comment (styfle) [reply package.json:16]: You're right. I just noticed its use for both `@vercel/next` and `@vercel/static-build` so lets save it for a future PR.

**Review by styfle — APPROVED** [review]
Great work, thanks! 🎉

**Review by javivelasco — APPROVED** [review]

**Pull request merged**
